src/coreclr/src/jit/earlyprop.cpp

```diff
@@ -644,41 +660,164 @@ void Compiler::optFoldNullCheck(GenTree* tree)
+            {
+                canRemoveNullCheck = false;
+            }
+            else
+            {
+                curStmt     = curStmt->GetPrevStmt();
+                currentTree = curStmt->GetRootNode();
+            }
+        }
+        *nullCheckStmt = curStmt;
+    }
+
+    if (canRemoveNullCheck && (*nullCheckParent == nullptr))
+    {
+        *nullCheckParent = nullCheckTree->gtGetParent(nullptr);
+    }
+
+    return canRemoveNullCheck;
+}
+
+//----------------------------------------------------------------
+// optCanMoveNullCheckPastTree: Check if a nullcheck node that is before `tree`
+//                              in execution order may be folded into an indirection node that
+//                              is after `tree` is execution order.
+//
+// Arguments:
+//    tree                  - The tree to check.
+//    nullCheckLclNum       - The local variable that GT_NULLCHECK checks.
+//    isInsideTry           - True if tree is inside try, false otherwise.
+//    checkExceptionSummary - If true, check GTF_EXCEPT flag only, otherwise check if the operation itself may throw.
 //
 // Return Value:
-//    True if GT_NULLCHECK can be folded into a node that is after tree is execution order,
+//    True if nullcheck may be folded into a node that is after tree is execution order,
 //    false otherwise.
 
-bool Compiler::optCanMoveNullCheckPastTree(GenTree* tree, bool isInsideTry)
+bool Compiler::optCanMoveNullCheckPastTree(GenTree* tree,
+                                           unsigned nullCheckLclNum,
+                                           bool     isInsideTry,
+                                           bool     checkExceptionSummary)
 {
-    bool result = true;
-    if (isInsideTry)
+    bool result = (tree->gtFlags & GTF_CALL) == 0;
+
+    if (result && (tree->gtFlags & GTF_EXCEPT) != 0)
     {
-        // We disallow calls, exception sources, and all assignments.
-        // Assignments to locals are disallowed inside try because
-        // they may be live in the handler.
-        if ((tree->gtFlags & GTF_SIDE_EFFECT) != 0)
-        {
-            result = false;
-        }
+        result = !checkExceptionSummary && !tree->OperMayThrow(this);
     }
-    else
+
+    if (result && ((tree->gtFlags & GTF_ASG) != 0))
     {
-        // We disallow calls, exception sources, and assignments to
-        // global memory.
-        if (GTF_GLOBALLY_VISIBLE_SIDE_EFFECTS(tree->gtFlags))
+        GenTree* lhs = tree->gtGetOp1();
+        if (isInsideTry)
         {
-            result = false;
+            // We disallow all assignments inside try except assignments to compiler temps.
+            // Assignments to user locals are disallowed because they may be live in the handler.
+            result = (lhs->OperGet() == GT_LCL_VAR) && lvaTable[lhs->AsLclVarCommon()->GetLclNum()].lvIsTemp &&
+                     ((tree->gtGetOp2()->gtFlags & GTF_ASG) == 0);
+        }
```

Comment:
Don't we know which locals are live into handlers by this point?

Reply:
Good point. I pushed a change to optCanMoveNullCheckPastTree that checks for lvLiveInOutOfHndlr and also fixes some issues with checking assignments. With this change I get additional improvements in frameworks with no new regressions:
```
Total bytes of diff: -7560 (-0.02% of base)
1538 total methods with Code Size differences (1526 improved, 12 regressed), 242942 unchanged.
```
and in benchmarks:
```
Total bytes of diff: -14 (-0.00% of base)
5 total methods with Code Size differences (5 improved, 0 regressed), 1888 unchanged.
```